File: src/main/java/mnm/mods/tabbychat/client/gui/component/layout/BorderLayout.java
package mnm.mods.tabbychat.client.gui.component.layout;

import com.google.common.collect.Maps;
import mnm.mods.tabbychat.client.gui.component.GuiComponent;
import mnm.mods.tabbychat.client.gui.component.GuiPanel;
import mnm.mods.tabbychat.util.Dim;
import mnm.mods.tabbychat.util.ILocation;
import mnm.mods.tabbychat.util.Location;

import java.util.EnumMap;
import java.util.Map.Entry;

/**
 * A recreation of Border Layout.
 *
 * @author Matthew Matthew
 * @see java.awt.BorderLayout
 */
public class BorderLayout implements ILayout {

    private EnumMap<Position, GuiComponent> components = Maps.newEnumMap(Position.class);

    @Override
    public void addComponent(GuiComponent comp, Object constraints) {
        if (constraints == null || constraints instanceof Position) {
            addComponent((Position) constraints, comp);
        } else {
            throw new IllegalArgumentException(
                    "cannot add to layout: constraint must be a Position enum");
        }
    }

    private synchronized void addComponent(Position constraint, GuiComponent comp) {
        if (constraint == null) {
            constraint = Position.CENTER;
        }

        components.put(constraint, comp);
    }

    @Override
    public synchronized void removeComponent(GuiComponent comp) {
        components.entrySet().stream()
                .filter(c -> c.getValue().equals(comp))
                .findFirst()
                .ifPresent(c -> components.remove(c.getKey()));
    }

    @Override
    public void layoutComponents(GuiPanel parent) {
        ILocation pbounds = parent.getLocation();
        GuiComponent center = components.get(Position.CENTER);
        GuiComponent north = components.get(Position.NORTH);
        GuiComponent south = components.get(Position.SOUTH);
        GuiComponent west = components.get(Position.WEST);
        GuiComponent east = components.get(Position.EAST);

        if (north != null) {
            north.setLocation(new Location(pbounds.getXPos(), pbounds.getYPos(), pbounds.getWidth(), north.getMinimumSize().height));
        }

        if (west != null) {
            Location bounds = pbounds.copy();
            bounds.setWidth(west.getMinimumSize().width);

            if (north != null) {
                bounds.setYPos(north.getLocation().getYHeight());
            }

            if (south == null) {
                if (north == null) {
                    bounds.setHeight(pbounds.getHeight());
                } else {
                    bounds.setHeight(pbounds.getHeight() - north.getLocation().getHeight());
                }
            } else {
                if (north == null) {
                    bounds.setHeight(pbounds.getHeight() - south.getLocation().getHeight());
                } else {
                    bounds.setHeight(pbounds.getHeight() - south.getLocation().getHeight() - north.getLocation().getHeight());
                }
            }
            west.setLocation(bounds);
        }

        if (center != null) {
            Location bounds = pbounds.copy();

            if (north != null) {
                bounds.setYPos(north.getLocation().getYHeight() + 1);
            }

            if (west != null) {
                bounds.setXPos(west.getLocation().getXWidth());
            }

            if (east == null) {
                if (west == null) {
                    bounds.setWidth(pbounds.getWidth());
                } else {
                    bounds.setWidth(pbounds.getWidth() - west.getLocation().getWidth());
                }
            } else {
                if (west == null) {
                    bounds.setWidth(pbounds.getWidth() - east.getLocation().getWidth());
                } else {
                    bounds.setWidth(pbounds.getWidth() - east.getLocation().getWidth() - west.getLocation().getWidth());
                }
            }

            if (south == null) {
                if (north == null) {
                    bounds.setHeight(pbounds.getHeight());
                } else {
                    bounds.setHeight(pbounds.getHeight() - north.getLocation().getHeight());
                }
            } else {
                if (north == null) {
                    bounds.setHeight(pbounds.getHeight() - south.getLocation().getHeight() - 1);
                } else {
                    bounds.setHeight(pbounds.getHeight() - south.getLocation().getHeight() - north.getLocation().getHeight() - 2);
                }
            }
            center.setLocation(bounds);
        }

        if (east != null) {
            Location bounds = pbounds.copy();

            bounds.setXPos(pbounds.getXWidth() - east.getMinimumSize().width);
            bounds.setWidth(east.getMinimumSize().width);
            if (north != null) {
                bounds.setYPos(north.getLocation().getYHeight());
            }
            if (south == null) {
                if (north == null) {
                    bounds.setHeight(pbounds.getHeight());
                } else {
                    bounds.setHeight(pbounds.getHeight() - north.getMinimumSize().height);
                }
            } else {
                if (north == null) {
                    bounds.setHeight(pbounds.getHeight() - south.getMinimumSize().height);
                } else {
                    bounds.setHeight(pbounds.getHeight() - south.getMinimumSize().height - north.getMinimumSize().height);
                }
            }
            east.setLocation(bounds);
        }

        if (south != null) {

            int x = pbounds.getXPos();
            int y = pbounds.getYHeight() - south.getLocation().getHeight();
            int width = pbounds.getWidth();
            int height = south.getMinimumSize().height;

            south.setLocation(new Location(x, y, width, height));
        }
    }

    @Override
    public Dim getLayoutSize() {
        int width = 0;
        int height = 0;
        for (Entry<Position, GuiComponent> comp : components.entrySet()) {
            Dim size = comp.getValue().getMinimumSize();
            switch (comp.getKey()) {
            case CENTER:
                width += size.width;
                height += size.height;
                break;
            case EAST:
            case WEST:
                width += size.width;
                break;
            case NORTH:
            case SOUTH:
                height += size.height;
                break;
            }
        }
        return new Dim(width, height);
    }

    public enum Position {
        NORTH,
        SOUTH,
        EAST,
        WEST,
        CENTER
    }
}

File: src/main/kotlin/mnm/mods/tabbychat/client/gui/component/layout/BorderLayout.kt
package mnm.mods.tabbychat.client.gui.component.layout

import mnm.mods.tabbychat.client.gui.component.GuiComponent
import mnm.mods.tabbychat.client.gui.component.GuiPanel
import mnm.mods.tabbychat.util.Dim
import mnm.mods.tabbychat.util.Location
import java.util.*

/**
 * A recreation of Border Layout.
 *
 * @author Matthew Matthew
 * @see java.awt.BorderLayout
 */
class BorderLayout : ILayout {

    private val components = EnumMap<Position, GuiComponent>(Position::class.java)

    override val layoutSize: Dim
        get() {
            var width = 0
            var height = 0
            for ((key, value) in components) {
                val (width1, height1) = value.minimumSize
                when (key!!) {
                    Position.CENTER -> {
                        width += width1
                        height += height1
                    }
                    Position.EAST, Position.WEST -> width += width1
                    Position.NORTH, Position.SOUTH -> height += height1
                }
            }
            return Dim(width, height)
        }

    override fun addComponent(comp: GuiComponent, constraints: Any?) {
        if (constraints == null || constraints is Position) {
            addComponent(constraints as Position?, comp)
        } else {
            throw IllegalArgumentException(
                    "cannot add to layout: constraint must be a Position enum")
        }
    }

    @Synchronized
    private fun addComponent(constraint: Position?, comp: GuiComponent) {
        components[constraint ?: Position.CENTER] = comp
    }

    @Synchronized
    override fun removeComponent(comp: GuiComponent) {
        components.values.remove(comp)
    }

    override fun layoutComponents(parent: GuiPanel) {
        val pbounds = parent.location
        val center = components[Position.CENTER]
        val north = components[Position.NORTH]
        val south = components[Position.SOUTH]
        val west = components[Position.WEST]
        val east = components[Position.EAST]

        if (north != null) {
            north.location = Location(pbounds.xPos, pbounds.yPos, pbounds.width, north.minimumSize.height)
        }

        if (west != null) {
            val bounds = pbounds.copy()
            bounds.width = west.minimumSize.width

            if (north != null) {
                bounds.yPos = north.location.yHeight
            }

            if (south == null) {
                if (north == null) {
                    bounds.height = pbounds.height
                } else {
                    bounds.height = pbounds.height - north.location.height
                }
            } else {
                if (north == null) {
                    bounds.height = pbounds.height - south.location.height
                } else {
                    bounds.height = pbounds.height - south.location.height - north.location.height
                }
            }
            west.location = bounds
        }

        if (center != null) {
            val bounds = pbounds.copy()

            if (north != null) {
                bounds.yPos = north.location.yHeight + 1
            }

            if (west != null) {
                bounds.xPos = west.location.xWidth
            }

            if (east == null) {
                if (west == null) {
                    bounds.width = pbounds.width
                } else {
                    bounds.width = pbounds.width - west.location.width
                }
            } else {
                if (west == null) {
                    bounds.width = pbounds.width - east.location.width
                } else {
                    bounds.width = pbounds.width - east.location.width - west.location.width
                }
            }

            if (south == null) {
                if (north == null) {
                    bounds.height = pbounds.height
                } else {
                    bounds.height = pbounds.height - north.location.height
                }
            } else {
                if (north == null) {
                    bounds.height = pbounds.height - south.location.height - 1
                } else {
                    bounds.height = pbounds.height - south.location.height - north.location.height - 2
                }
            }
            center.location = bounds
        }

        if (east != null) {
            val bounds = pbounds.copy()

            bounds.xPos = pbounds.xWidth - east.minimumSize.width
            bounds.width = east.minimumSize.width
            if (north != null) {
                bounds.yPos = north.location.yHeight
            }
            if (south == null) {
                if (north == null) {
                    bounds.height = pbounds.height
                } else {
                    bounds.height = pbounds.height - north.minimumSize.height
                }
            } else {
                if (north == null) {
                    bounds.height = pbounds.height - south.minimumSize.height
                } else {
                    bounds.height = pbounds.height - south.minimumSize.height - north.minimumSize.height
                }
            }
            east.location = bounds
        }

        if (south != null) {

            val x = pbounds.xPos
            val y = pbounds.yHeight - south.location.height
            val width = pbounds.width
            val height = south.minimumSize.height

            south.location = Location(x, y, width, height)
        }
    }

    enum class Position {
        NORTH,
        SOUTH,
        EAST,
        WEST,
        CENTER
    }
}
